Orange/widgets/model/owneuralnetwork.py

```diff
@@ -81,6 +124,118 @@ def get_hidden_layers(self):
             self.hidden_layers_edit.setText("100,")
         return layers
 
+    def update_model(self):
+        self.show_fitting_failed(None)
+        self.model = None
+        if self.check_data():
+            self.__update()
+        else:
+            self.Outputs.model.send(self.model)
+
+    @Slot(float)
+    def setProgressValue(self, value):
+        assert self.thread() is QThread.currentThread()
+        self.progressBarSet(value)
+
+    def __update(self):
+        if self._task is not None:
+            # First make sure any pending tasks are cancelled.
+            self.cancel()
+        assert self._task is None
+
+        self.setBlocking(True)
+
+        self._task = task = Task()
+
+        # A thread safe way to invoke a method
+        set_progress = methodinvoke(self, "setProgressValue", (float,))
+
+        max_iter = self.learner.kwargs["max_iter"]
+
+        def callback(iteration=None):
+            if task.cancelled:
+                raise CancelThreadException()  # this stop the thread
+            if iteration is not None:
+                set_progress(iteration/max_iter*100)
+
+        def print_callback(*args, **kwargs):
+            iters = None
+            # try to parse iteration number
+            if args and args[0] and isinstance(args[0], str):
+                find = re.findall(r"Iteration (\d+)", args[0])
+                if find:
+                    iters = int(find[0])
+            callback(iters)
+
+        def build_model(data, learner):
+            if learner.kwargs["solver"] != "lbfgs":
+                # enable verbose printouts within scikit and redirect them
+                with patch.dict(learner.kwargs, {"verbose": True}),\
+                     patch("builtins.print", print_callback):
```

Comment:
```
Time -->

Thread-1
-|patch|-----|exit|--------

Thread-2
----|patch|--------|exit|--
          ^             ^
         stores         restores
     patched print      the patched print of Thread-1

```

Reply:
Whoops. Thank you. Did not know that patch mutates state globally....